compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt

```diff
@@ -0,0 +1,212 @@
+
+                    Action.DISPOSE -> {
+                        logWarning("'ViewControllerBasedLifecycleOwner' received 'Action.DISPOSE' while in 'State.Running'. Make sure that view controller containment API is used correctly. 'removeFromParent' must be called before 'dispose'")
+
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+
+                        Disposed
+                    }
+                }
+            }
+        }
+
+        class Suspended(private val lifecycle: LifecycleRegistry) : State {
+            override fun reduce(action: Action): State {
+                return when(action) {
+                    Action.VIEW_WILL_APPEAR -> {
+                        this
+                    }
+
+                    Action.VIEW_DID_DISAPPEAR -> {
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
+
+                        Created(isApplicationForeground = false, lifecycle = lifecycle)
+                    }
+
+                    Action.APPLICATION_DID_ENTER_BACKGROUND -> {
+                        this
+                    }
+
+                    Action.APPLICATION_WILL_ENTER_FOREGROUND -> {
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
+
+                        Running(lifecycle = lifecycle)
+                    }
+
+                    Action.DISPOSE -> {
+                        logWarning("'ViewControllerBasedLifecycleOwner' received 'Action.DISPOSE' while in 'State.Suspended'. Make sure that view controller containment API is used correctly. 'removeFromParent' must be called before 'dispose'")
+
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
+                        lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+
+                        Disposed
+                    }
+                }
+            }
+        }
+
+        object Disposed : State {
+            override fun reduce(action: Action): State {
+                when (action) {
+                    Action.VIEW_WILL_APPEAR, Action.VIEW_DID_DISAPPEAR, Action.DISPOSE -> {
+                        fail("Invalid '$action' for 'State.Disposed'")
+                    }
+                    Action.APPLICATION_DID_ENTER_BACKGROUND, Action.APPLICATION_WILL_ENTER_FOREGROUND -> {
+                        // no-op
+                        return this
+                    }
+                }
+            }
+        }
+    }
+
+    override val lifecycle = LifecycleRegistry(this)
+
+    private var state: State = State.Created(
+        isApplicationForeground = ApplicationStateListener.isApplicationActive,
+        lifecycle = lifecycle
+    )
+
+    private val applicationStateListener = ApplicationStateListener { isForeground ->
+        state = state.reduce(
```

Comment:
It's better to call `this.reduce()` here to have single point of state changes.

Reply:
That's indeed an oversight, fixed